rfcs/003-strings.md

```diff
@@ -0,0 +1,325 @@
+    fn as_wide(&self) -> &[u16] {}
+    // String data with the trailing 0
+    fn as_wide_with_nul(&self) -> &[u16] {}
+    fn as_pcwstr(&self) -> PCWSTR {}
+    fn to_string(&self) -> Result<String, FromUtf16Error> {}
+    fn to_string_lossy(&self) -> String {}
+}
+
+// This allows `CWString` to be turned into a `Borrowed<'a, PCWSTR>`
+impl From<&CWString> for &PCWSTR {}
+impl From<CWString> for HSTRING {}
+impl From<HSTRING> for CWString {}
+
+impl Drop for CWString {
+    // The `CWString` owns its memory and will free when its dropped
+}
+// Display shows the string with non-valid utf16 replaced with �
+impl Display for CWSTRING {}
+// Same as Display but surrounded by ""
+impl Debug for CWSTRING {}
+
+// HSTRING should also include implementations of `PartialEq` for 
+// `String`, `&str`, `OsString`, and `&OsStr`
+
+// --------------------------------
+
+impl BSTR {
+    fn from_ptr(ptr: *const c_void) -> Self {}
+    fn as_ptr(&self) -> *const c_void;
+    // String data without the trailing 0
+    unsafe fn as_wide(&self) -> &[u16] {}
+    // String data with the trailing 0
+    unsafe fn as_wide_with_nul(&self) -> &[u16] {}
+    unsafe fn to_string(&self) -> Result<String, FromUtf16Error> {}
+    unsafe fn to_string_lossy(&self) -> String {}
+}
+
+
+// --------------------------------
+
+impl PCWSTR {
+    fn from_ptr(ptr: *const u16) -> Self {}
+    fn as_ptr(&self) -> *const u16;
+    fn is_null(&self) -> bool {}
+    // String data without the trailing 0
+    unsafe fn as_wide(&self) -> &[u16] {}
+    // String data with the trailing 0
+    unsafe fn as_wide_with_nul(&self) -> &[u16] {}
+    unsafe fn to_string(&self) -> Result<String, FromUtf16Error> {}
+    unsafe fn to_string_lossy(&self) -> String {}
+    unsafe fn display(&self) -> impl Display {}
+}
+// This allows `CString` and `CSTR` to be turned into a `Borrowed<'a, PCSTR>`
+impl From<&CString> for &PCSTR {}
+impl From<&CStr> for &PCSTR {}
+
+// --------------------------------
+
+impl PCSTR {
+    fn from_ptr(ptr: *const u8) -> Self {}
+    fn as_ptr(&self) -> *const u8;
+    fn is_null(&self) -> bool {}
+    // String data without the trailing 0
+    unsafe fn as_bytes(&self) -> &[u8] {}
+    // String data with the trailing 0
+    unsafe fn as_bytes_with_nul(&self) -> &[u8] {}
+    // Converts to `&str` checking for valid UTF-8
+    unsafe fn as_str(&self) -> Result<&str, FromUtf8Error> {}
+    // Converts to `&str` not checking for valid UTF-8
+    unsafe fn as_str_unchecked(&self) -> &str {}
+}
+
+// Note that `Display`, `Debug`, and the conversion traits are not 
+// included for non-`HSTRING` types because those traits are not 
+// marked as unsafe
+```
+
+#### String literals
+
+Many times the user simply wants to use a string literal to build a string of a certain type.
+
+The windows crate provides the following macros for convenience:
+
+```rust
+let x: CWString = w!("hello");
+let y: CString = c!("hello");
+let z: HSTRING = h!("hello");
```

Comment:
I don't think we need these macros to produce allocated strings. That requires runtime support. They should all be able to return `const` string literals of `PCWSTR` and `PCSTR` respectively. Then they can simply be used to call functions like `MessageBoxA/W` directly and efficiently.

Reply:
Common Windows string types like `HSTRING` and `BSTR` can then just provide `From<&T>` implementations for `PCWSTR` for cases where a string needs to be computed at runtime and in future if/when `CString` and `CWString` are available in the std library we can just provide `From` implementations for those as well.

Reply:
Then we shouldn't need the `Into<T>` below, unless I'm forgetting something.

Reply:
Here's what I have in mind: https://github.com/microsoft/windows-rs/compare/utf16?expand=1

Reply:
And this is compatible with the existing state of the `windows` crate so would make it easier to land the `Borrowed` PR.

Reply:
This proposal is a subset of what I had in mind so I think it's fine to try it out. We can always expand from there to have dedicated `CWString` types and such.